### PR TITLE
Feature/29 export import

### PR DESCRIPTION
**Pull request by JonasJurczok**

### CODE REVIEW

**Comment by JonasJurczok** [comment]
@Tarrke @hoylemd please take a look. Especially how the sprite button is rendered. I'm not sure if it works the way it is supposed to work as I have no experience with sprites/icons.

File: README.md [hunk 1]
```diff
@@ -73,19 +73,30 @@ an idea for a feature.
...
+* It is now possible to delete tasks
+* It is now possible to export and import tasks similar to blueprints
 * The button text now switches between the following states
   * No tasks available: Todo List
   * Tasks available: Todo List: X tasks available
   * Task assigned to player: Todo List: <first line of task>
-* It is now possible to delete tasks
 * Fixed credits for 16.4.0
 
+### 16.5.1
```
Comment (Tarrke) [inline README.md]: There is no 16.5.0 version ?

File: locale/fr/locale.cfg [hunk 2]
```diff
@@ -14,15 +14,18 @@ add_top=[Add to Top]
 persist=Enregistrer
 cancel=Annuler
 update=MAJ
-delete=Delete
-confirm_deletion=Confirm deletion
+delete=Supprimer
+confirm_deletion=Confirmer la suppression
+
+import=import
```
Comment (Tarrke) [inline locale/fr/locale.cfg]:
```suggestion
import=importer
```

File: locale/fr/locale.cfg [hunk 3]
```diff
@@ -14,15 +14,18 @@ add_top=[Add to Top]
...
+export=export
```
Comment (Tarrke) [inline locale/fr/locale.cfg]:
```suggestion
export=exporter
```
This one, as the one above may have to be corrected, I have to check the whole string and sentences where it is used. Should be correct if only used as a label.

File: spec/29_import_export_feature.lua [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+
+        assert(button.enabled == false)
+
+    end)
+
+    scenario("Adding/removing tasks while export dialog open is ignored.", function()
+        -- we purposely ignore this problem as it is very complex to solve and an extreme edge case
+        assert(true == true)
+    end)
+
```
Comment (Tarrke) [inline spec/29_import_export_feature.lua]: Should have:
* scenario "Export done tasks is possible"

File: src/todo/UI.lua [hunk 5]
```diff
@@ -132,6 +150,127 @@ function todo.create_task_table(frame, player)
     return table
 end
 
+function todo.create_export_dialog(player)
```
Comment (Tarrke) [inline src/todo/UI.lua]:
```suggestion
function todo.create_export_frame(player)
```

Here you use dialog, the rest of the code use frame for the same thing. Maybe we should keep to frame or change everthing else to dialog. IMO both are good description for the pop-up window.

File: src/todo/UI.lua [hunk 6]
```diff
@@ -132,6 +150,127 @@ function todo.create_task_table(frame, player)
...
+    flow.add({
+        type = "button",
+        style = "todo_button_default",
+        name = "todo_export_button",
+        caption = {"todo.export"}
+    })
+
+end
+
+function todo.create_import_dialog(player)
```
Comment (Tarrke) [inline src/todo/UI.lua]:
```suggestion
function todo.create_import_frame(player)
```
See above

File: src/todo/UI.lua [hunk 7]
```diff
@@ -132,6 +150,127 @@ function todo.create_task_table(frame, player)
     return table
 end
 
+function todo.create_export_dialog(player)
+    local gui = player.gui.center
+
+    if (gui.todo_export_dialog ~= nil) then
```
Comment (Tarrke) [inline src/todo/UI.lua]:
```suggestion
    if (gui.todo_export_frame ~= nil) then
```
See above

File: src/todo/UI.lua [hunk 8]
```diff
@@ -132,6 +150,127 @@ function todo.create_task_table(frame, player)
...
+        gui.todo_export_dialog.destroy()
```
Comment (Tarrke) [inline src/todo/UI.lua]:
```suggestion
        gui.todo_export_frame.destroy()
```
See above

File: src/todo/UI.lua [hunk 9]
```diff
@@ -132,6 +150,127 @@ function todo.create_task_table(frame, player)
...
+    end
+
+    local frame = gui.add({
+        type = "frame",
+        name = "todo_export_dialog",
```
Comment (Tarrke) [inline src/todo/UI.lua]:
```suggestion
        name = "todo_export_frame",
```
See above

File: src/todo/UI.lua [hunk 10]
```diff
@@ -132,6 +150,127 @@ function todo.create_task_table(frame, player)
...
+        name = "todo_export_button",
+        caption = {"todo.export"}
+    })
+
+end
+
+function todo.create_import_dialog(player)
+    local gui = player.gui.center
+
+    if (gui.todo_import_dialog ~= nil) then
```
Comment (Tarrke) [inline src/todo/UI.lua]:
```suggestion
    if (gui.todo_import_frame ~= nil) then
```
See above

File: src/todo/UI.lua [hunk 11]
```diff
@@ -132,6 +150,127 @@ function todo.create_task_table(frame, player)
...
+        gui.todo_import_dialog.destroy()
```
Comment (Tarrke) [inline src/todo/UI.lua]:
```suggestion
        gui.todo_import_frame.destroy()
```
See above

File: src/todo/UI.lua [hunk 12]
```diff
@@ -132,6 +150,127 @@ function todo.create_task_table(frame, player)
...
+    end
+
+    local frame = gui.add({
+        type = "frame",
+        name = "todo_import_dialog",
```
Comment (Tarrke) [inline src/todo/UI.lua]:
```suggestion
        name = "todo_import_frame",
```
See above

File: src/todo/helper.lua [hunk 13]
```diff
@@ -75,6 +74,13 @@ function todo.get_task_table(player)
     end
 end
 
+function todo.get_export_dialog(player)
+    local center = player.gui.center
+    if (center.todo_export_dialog) then
```
Comment (Tarrke) [inline src/todo/helper.lua]:
```suggestion
    if (center.todo_export_frame) then
```
See above

File: src/todo/helper.lua [hunk 14]
```diff
@@ -75,6 +74,13 @@ function todo.get_task_table(player)
...
+        return center.todo_export_dialog
```
Comment (Tarrke) [inline src/todo/helper.lua]:
```suggestion
        return center.todo_export_frame
```
See above

File: src/todo/todo.lua [hunk 15]
```diff
@@ -65,9 +67,84 @@ function todo.save_task(task, should_add_to_top)
 
     table.insert(global.todo.open, add_index, task)
 
+    todo.update_export_dialog_button_state()
```
Comment (Tarrke) [inline src/todo/todo.lua]:
```suggestion
    todo.update_export_frame_button_state()
```
See above

File: src/todo/todo.lua [hunk 16]
```diff
@@ -65,9 +67,84 @@ function todo.save_task(task, should_add_to_top)
...
+
     return task
 end
 
+function todo.update_export_dialog_button_state()
```
Comment (Tarrke) [inline src/todo/todo.lua]:
```suggestion
function todo.update_export_frame_button_state()
```
See above

File: src/todo/todo.lua [hunk 17]
```diff
@@ -65,9 +67,84 @@ function todo.save_task(task, should_add_to_top)
...
+    local task_count = #global.todo.open + #global.todo.done
+
+    for _, player in pairs(game.players) do
+        local main_frame = todo.get_main_frame(player)
```
Comment (Tarrke) [inline src/todo/todo.lua]:
```suggestion
        local player_main_frame = todo.get_main_frame(player)
```
This variable name should reflect the fact that this is the main frame a of a player, not a local main_frame.

File: src/todo/todo.lua [hunk 18]
```diff
@@ -65,9 +67,84 @@ function todo.save_task(task, should_add_to_top)
...
+        if (main_frame) then
+            main_frame.todo_main_button_flow.todo_export_dialog_button.enabled = task_count > 0
+        end
+
+    end
+end
+
+function todo.generate_and_show_export_string(player)
+    local dialog = todo.get_export_dialog(player)
```
Comment (Tarrke) [inline src/todo/todo.lua]:
```suggestion
    local frame = todo.get_export_frame(player)
```
See above

File: src/todo/todo.lua [hunk 19]
```diff
@@ -65,9 +67,84 @@ function todo.save_task(task, should_add_to_top)
...
+
+    local tasks_table = dialog.todo_export_dialog_scroll_pane.todo_export_dialog_table
```
Comment (Tarrke) [inline src/todo/todo.lua]:
```suggestion
    local tasks_table = frame.todo_export_frame_scroll_pane.todo_export_frame_table
```
Still see above

File: src/todo/todo.lua [hunk 20]
```diff
@@ -333,6 +410,7 @@ function todo.on_gui_click(event)
         todo.toggle_show_completed(player)
         todo.refresh_task_table(player)
     elseif (element.name == "todo_cancel_button") then
+        -- close the edit dialog
```
Comment (Tarrke) [inline src/todo/todo.lua]: Thanks!

File: README.md (hunk 1 above)
Comment (JonasJurczok) [reply README.md]: I fucked up the versions when doing the crash fix a couple days ago..
I will have to develop a more.. stable versioning regiment in the future -.-

File: locale/fr/locale.cfg (hunk 3 above)
Comment (JonasJurczok) [reply locale/fr/locale.cfg]: So far it is only used as label :) Thanks

File: spec/29_import_export_feature.lua (hunk 4 above)
Comment (JonasJurczok) [reply spec/29_import_export_feature.lua]: true

File: src/todo/UI.lua (hunk 5 above)
Comment (JonasJurczok) [reply src/todo/UI.lua]: Yeah I also realised this is inconsistent..
My thinking was "Frame" is something like the main frame and other static frames.
Dialogs are popups that are only used for a short amount of time and block gui.center.

I thought about doing another major rework of everything including abstracting all this manual "lets create the same dialog stub over and over again" away. Then we would also have a more consistent naming.

File: src/todo/todo.lua (hunk 20 above)
Comment (JonasJurczok) [reply src/todo/todo.lua]: :D

**Comment by JonasJurczok** [comment]
Thanks for the comments.
As I wrote above I would like to rework the inner workings of the mod another time as it is quite spagetti by now..

Regarding the attributions: there is an attributions section above the version history now. This contains the links to the libs as well.

**Pull request merged**